File: app/src/main/java/com/moufee/purduemenus/db/FavoriteDao.java
package com.moufee.purduemenus.db;

import com.moufee.purduemenus.menus.Favorite;

import java.util.List;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;

@Dao
public interface FavoriteDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    public void insertFavorites(List<Favorite> favorites);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    public void insertFavorites(Favorite... favorites);

    @Query("SELECT * FROM favorite")
    public LiveData<List<Favorite>> loadAllFavorites();

    @Query("SELECT * FROM favorite")
    public List<Favorite> getAllFavorites();

    @Query("SELECT itemId FROM favorite")
    public LiveData<List<String>> getFavoriteIDs();

    @Query("SELECT * FROM favorite WHERE itemId = :itemID LIMIT 1")
    public Favorite getFavoriteByItemId(String itemID);

    @Query("DELETE FROM favorite WHERE itemId = :itemID")
    public void deleteByItemID(String itemID);

    @Delete
    public void deleteFavorites(Favorite... favorites);

    @Query("DELETE FROM favorite")
    public void deleteAll();
}

File: app/src/main/java/com/moufee/purduemenus/db/FavoriteDao.kt
package com.moufee.purduemenus.db

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import com.moufee.purduemenus.repository.data.menus.Favorite
import kotlinx.coroutines.flow.Flow

@Dao interface FavoriteDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE) fun insertFavorites(favorites: List<Favorite>)

    @Insert(onConflict = OnConflictStrategy.REPLACE) suspend fun insertFavorites(vararg favorites: Favorite)

    @Query("SELECT * FROM favorite") fun getFavoritesFlow(): Flow<List<Favorite>>

    @Query("SELECT * FROM favorite") suspend fun getAllFavorites(): List<Favorite>

    @Query("SELECT itemId FROM favorite") fun getFavoriteIDs(): Flow<List<String>>

    @Query("SELECT * FROM favorite WHERE itemId = :itemID LIMIT 1") suspend fun getFavoriteByItemId(itemID: String): Favorite?

    @Query("DELETE FROM favorite WHERE itemId = :itemID") suspend fun deleteByItemID(itemID: String)

    @Query("DELETE FROM favorite") suspend fun deleteAll()
}